cmd/acr-cleanup/main.go

```diff
@@ -0,0 +1,356 @@
+	http.Handle("/metrics", promhttp.Handler())
+	log.Fatal(http.ListenAndServe(":8080", nil))
+}
+
+func maintainImages(radixClient radixclient.Interface, cleanupDays []string,
+	cleanupStart string, cleanupEnd string, period time.Duration,
+	registry, clusterType string, deleteUntagged bool,
+	retainLatestUntagged int, performDelete bool, whitelisted []string) {
+	window, err := timewindow.New(cleanupDays, cleanupStart, cleanupEnd, timezone)
+	if err != nil {
+		log.Fatalf("Failed to build time window: %v", err)
+	}
+
+	source := rand.NewSource(time.Now().UnixNano())
+	tick := delaytick.New(source, period)
+	for time := range tick {
+		if window.Contains(time) {
+			log.Infof("Start deleting images %s", time)
+			deleteImagesBelongingTo(radixClient, registry, clusterType, deleteUntagged, retainLatestUntagged, performDelete, whitelisted)
+		} else {
+			log.Infof("%s is outside of window. Continue sleeping", time)
+		}
+	}
+}
+
+func initializeFlagSet() *pflag.FlagSet {
+	// Flag domain.
+	fs := pflag.NewFlagSet("default", pflag.ContinueOnError)
+	fs.Usage = func() {
+		fmt.Fprintf(os.Stderr, "DESCRIPTION\n")
+		fmt.Fprintf(os.Stderr, "  Radix acr cleanup.\n")
+		fmt.Fprintf(os.Stderr, "\n")
+		fmt.Fprintf(os.Stderr, "FLAGS\n")
+		fs.PrintDefaults()
+	}
+	return fs
+}
+
+func parseFlagsFromArgs(fs *pflag.FlagSet) {
+	err := fs.Parse(os.Args[1:])
+	switch {
+	case err == pflag.ErrHelp:
+		os.Exit(0)
+	case err != nil:
+		fmt.Fprintf(os.Stderr, "Error: %s\n\n", err.Error())
+		fs.Usage()
+		os.Exit(2)
+	}
+}
+
+func deleteImagesBelongingTo(radixClient radixclient.Interface, registry, clusterType string,
+	deleteUntagged bool, retainLatestUntagged int, performDelete bool, whitelisted []string) {
+	start := time.Now()
+
+	defer func() {
+		duration := time.Since(start)
+		log.Infof("It took %s to run", duration)
+	}()
+
+	imagesInCluster, err := listActiveImagesInCluster(radixClient)
+	if err != nil {
+		log.Fatalf("Unable to list images in cluster, %v. Cannot proceed", err)
+	}
+
+	repositories := acr.ListRepositories(registry)
+
+	numRepositories := len(repositories)
+	processedRepositories := 0
+
+	for _, repository := range repositories {
+		if isWhitelisted(repository, whitelisted) {
+			log.Infof("Skip repository %s, as it is whitelisted", repository)
+			continue
+		}
+
+		log.Debugf("Process repository %s", repository)
+		manifests := acr.ListManifests(registry, repository)
+		numManifests := len(manifests)
+
+		for _, manifest := range manifests {
+			isNotTaggedForAnyClustertype := manifest.IsNotTaggedForAnyClustertype()
+
+			// If this manifest has a timestamp newer than start,
+			// the list of images might not be correct
+			if manifest.Timestamp.After(start) {
+				if isNotTaggedForAnyClustertype {
+					addUntaggedImageRetained(clusterType, repository)
+				} else {
+					addImageRetained(clusterType, repository)
+				}
+
+				continue
+			}
+
+			manifestExistInCluster := manifestExistInCluster(repository, manifest, imagesInCluster)
```

Comment:
Avoid naming variable exactly the same as function name (maybe add _is_ in front of it?).

Reply:
Done